### DISCUSSION TITLE
Should one use 3D Slicer on a remote server or simply use the OHIF option?

### DISCUSSION THREAD

**Discussion by Zrrr1997** [Project-MONAI/MONAILabel#776] [General]
Dear MONAILabel team,

thanks for the great project, the more I read about it and use it, the more I realize how awesome it is!
I have been playing around with the tutorials and with the MSD datasets to get to know the basic functionality. 

I was wondering whether it is possible to run 3D Slicer with the MONAILabel extension on a remote server and **interact** with the GUI the same way one would on a local machine. Currently, I have only tested this with X11 forwarding, which is very dependent on the bandwidth and leads to high latency feedback from the remote server. Another option I have considered is using [SlicerJupyter](https://github.com/Slicer/SlicerJupyter), but this limits the interaction even more.

3D Slicer offers a very intuitive interface for interaction but it lacks remote server support. Would it be more realistic to simply use the OHIF option or build your own viewer with the REST-ful API? Or are there some options with Slicer's interface which already support this functionality?

Best,
Zdravko

**Comment by SachidanandAlle** [comment]
Push your images some dicom Web (Orthanc) and use ohif viewer which is as part of monailabel server

In later stages depending on your use-case you can directly consume rest apis and integrate the functionality in your own client.. what we provided is only few reference implementation :)

**Comment by diazandr3s** [comment]
Hi @Zrrr1997,

Thanks for opening this discussion. We're happy you find MONAI Label useful :)

> I was wondering whether it is possible to run 3D Slicer with the MONAILabel extension on a remote server and interact with the GUI the same way one would on a local machine. Currently, I have only tested this with X11 forwarding, which is very dependent on the bandwidth and leads to high latency feedback from the remote server. Another option I have considered is using [SlicerJupyter](https://github.com/Slicer/SlicerJupyter), but this limits the interaction even more.

You could start the MONAI Label server on a GPU-based machine and then connect to it using Slicer. For that, you need to make sure the port is opened (i.e. port 8000) and the firewall allows you to connect from the internet. Is that possible in your step up?

> 3D Slicer offers a very intuitive interface for interaction but it lacks remote server support. Would it be more realistic to simply use the OHIF option or build your own viewer with the REST-ful API? Or are there some options with Slicer's interface which already support this functionality?

Another option is OHIF, for that you don't need to install anything. However, you don't have all the manual annotation tools available in Slicer. 

In order to use OHIF, you should have the DICOMWeb server (i.e. [Orthanc](https://github.com/Project-MONAI/MONAILabel/tree/main/plugins/ohif#installing-orthanc-dicomweb)) up and running as well.

**Reply by Zrrr1997** [reply]
Actually, as Andres stated, the answer is quite simple - just run the MONAI Label server on the remote server and Slicer on your local machine with port forwarding. So while Slicer itself does not support remote access, its **plugin** for MONAI Label does support this. I was so caught up with Slicer's functionality that I overlooked that MONAI already supports this.

I actually got it working and will post the (very simple) solution here if anyone else finds himself wondering like me. You simply need to forward any port you wish (for example `9090`) from your remote server e.g.:

`ssh -L 9090:localhost:8000 username@remote-server` 

This would create an SSH tunnel from your local port 9090 to the remote server's port 8000 (default port for MONAI Label server). Then you would simply need to put `http://localhost:9090/` in 3D Slicer's `MONAI Label server` field and you would get your latency-free Slicer viewer.

Thanks for clearing things up! Keep up the good work :)

**Reply by lassoan** [reply]
> I was wondering whether it is possible to run 3D Slicer with the MONAILabel extension on a remote server and interact with the GUI the same way one would on a local machine.

This is what MONAILabelReviewer project people do: use a single shared remote MONAILabel server and many people connect to it from their locally installed Slicer. It works well both within the hospital and remotely via VPN.

> 3D Slicer offers a very intuitive interface for interaction but it lacks remote server support

Slicer has extensive remote server support. Both classis DICOM storage, push, query/retrieve and DICOMweb. You can also launch a local Slicer instance directly from a webpage, etc. See a few example use cases in this demo: https://youtu.be/60_hzSlptWY

You can also use Slicer and OHIF from the same database, so you can generate content in Slicer and view in OHIF and vice versa.

**Reply by Zrrr1997** [reply]
This is great! MONAILabelReviewer seems like the perfect quality assurance project for high-quality labels. It seems that a local Slicer + remote MONAILabel is the standard way to go. I would imagine that this makes sense as most often you would want to use GPUs on the MONAILabel side for the training/inference of the segmentation models. 

I will read up more about Slicer as it seems I have missed a large portion of its functionality. However, I am getting the big picture now about remote processing thanks to this great discussion!
